src/base/hip/hip_conversion.cpp

```diff
@@ -637,6 +638,162 @@ namespace rocalution
+
+            // transpose matrix so that dst values are row major
+            rocblas_status blas_status = rocblasTgeam(blas_handle,
+                                                      rocblas_operation_transpose,
+                                                      rocblas_operation_none,
+                                                      nrow,
+                                                      ncol,
+                                                      &alpha,
+                                                      temp,
+                                                      nrow,
+                                                      &beta,
+                                                      B,
+                                                      nrow,
+                                                      dst->val,
+                                                      nrow);
+            CHECK_ROCBLAS_ERROR(blas_status, __FILE__, __LINE__);
+
+            free_hip(&temp);
+        }
+
+        // Sync memcopy
+        hipDeviceSynchronize();
+
+        return true;
+    }
+
+    template <typename ValueType, typename IndexType>
+    bool dense_to_csr_hip(const rocsparse_handle           sparse_handle,
+                          const rocblas_handle             blas_handle,
+                          IndexType                        nrow,
+                          IndexType                        ncol,
+                          const MatrixDENSE<ValueType>&    src,
+                          MatrixCSR<ValueType, IndexType>* dst,
+                          const rocsparse_mat_descr        dst_descr)
+    {
+        assert(nrow > 0);
+        assert(ncol > 0);
+
+        assert(dst != NULL);
+        assert(sparse_handle != NULL);
+        assert(blas_handle != NULL);
+        assert(dst_descr != NULL);
+
+        IndexType nnz_total;
+        IndexType* nnz_per_row = NULL;
+
+        if(DENSE_IND_BASE == 0)
+        {
+            rocsparse_status status;
+
+            allocate_hip(nrow, &nnz_per_row);
+
+            status = rocsparseTnnz(
+                sparse_handle, rocsparse_direction_row, nrow, ncol, dst_descr, src.val, nrow, nnz_per_row, &nnz_total);
+            CHECK_ROCSPARSE_ERROR(status, __FILE__, __LINE__);  
+
+            allocate_hip(nrow + 1, &dst->row_offset);
+            allocate_hip(nnz_total, &dst->col);
+            allocate_hip(nnz_total, &dst->val);
+
+            status = rocsparseTdense2csr(
+                sparse_handle, nrow, ncol, dst_descr, src.val, nrow, nnz_per_row, dst->val, dst->row_offset, dst->col);
+            CHECK_ROCSPARSE_ERROR(status, __FILE__, __LINE__);
+
+            free_hip(&nnz_per_row);
+        }
+        else
+        {
+            ValueType* temp = NULL;
+            allocate_hip(nrow * ncol, &temp);
+
+            ValueType alpha = static_cast<ValueType>(1);
+            ValueType beta = static_cast<ValueType>(0);
+
+            ValueType* B; 
```

Comment:
Doesn't B need to be allocated by the user?

Reply:
See above